projecttacoma/deqm-test-server-frontend
# Update resource implementation

## Summary
Implementation of the update resource button. Loads a ResourceCodeEditor component pre-filled with the resource’s JSON body. A PUT request with the code editor's contents is sent to the server when the submit button is clicked.

## New behavior
In any /[resourceType]/[id], there is now an Update button. When clicked, a new page is rendered with a ResourceCodeEditor component and an “Update Resource” button. The code editor is pre-filled with the resource’s body. The contents can be edited, then when the “Update Resource” button is clicked, a PUT request with the code editor’s contents is sent to the test server.  
- If a resource is successfully updated:

  1.    a **success notification** appears 

  1.   and the user is **redirected** to the resource's home page, where the updated body is displayed.

- If the request is unsuccessful (server returns 401, fetch throws an error, etc.) an error message notification appears on the update resource page.
 
## Code changes
- **pages/[resourceType]/[id]** folder added, includes index.tsx to replace [id].tsx
- **pages/[resourceType]/[id]/update.tsx**  added, implements update resource functionality
-  **pages/[resourceType]/[id]/index.tsx** now includes update button
- **__tests__/helpers/testHelpers.tsx** mockFetchImplementation function customized, new function added for testing with CodeMirror
- **__tests__/pages/resource/id/update.test.tsx** added to test the new update functionality
- **__tests__/pages/resource/id/index.test.tsx** expects update link button
- tests folder renaming files and some restructuring

## Testing Guidance
- See the README.md for first time setup instructions.
- To start the frontend: `npm run dev` then navigate to http://localhost:3000 in a browser
- Click on any resource type from the navigation menu on the left. Then click on any of the listed resource IDs. This will load that resource's home. Then click on the Update button in the upper right hand corner of the main page body. This should load a new page with a ResourceCodeEditor component and an “Update Resource” button
Test the update implementation from here in the following 3 ways:

  1. Click the “Update Resource” button. User should be redirected to the resource’s page displaying its JSON body. A green check-marked success notification should appear.
  2. Edit the resource’s ID in the code editor. Then click the “Update Resource” button. A red X-marked error notification will appear with a 40x error message.
  3. Disconnect from the server, then click the “Update Resource” button. A red X-marked error notification will appear with a "Problem connecting to server" message.

- Run the unit tests: `npm run test`
    - Note that one test in tests/pages/resource/create.test.tsx is skipped. This is unrelated to the functionality of this branch.
